src/rpc/core_rpc_server.cpp

```diff
@@ -919,6 +1021,11 @@ namespace cryptonote
   bool core_rpc_server::on_getblockcount(const COMMAND_RPC_GETBLOCKCOUNT::request& req, COMMAND_RPC_GETBLOCKCOUNT::response& res)
   {
     PERF_TIMER(on_getblockcount);
+    if (m_should_use_bootstrap_daemon)
```

Comment:
This is still a problem and undefined behavior as `m_should_use_bootstrap_daemon` may be written to in another thread. You need to lock the mutex here, too.

Reply:
I don’t think so, the race condition can occur only when multiple threads try to wtite to the same data simultaneously. Reading only doesn’t require locking.

Reply:
Having multiple _readers_ is not a problem, but as soon as there's an additional concurrent writer ([this one](https://github.com/monero-project/monero/pull/3165/files#diff-4da80f3ef9fb0e5910c35774e915537cR1260)), it results in undefined behavior.

Data race tl;dr:

1. __Multiple concurrent readers__: Not a problem
2. __Multiple concurrent writers__: Undefined behavior
3. __A writer and a concurrent reader__: Undefined behavior

Reply:
There **is** a lock, don’t you see it? Or am I mistaken?

Reply:
You need to lock the mutex before _reading from_ __and__ _writing to_ a shared variable.

Reply:
No, you’re wrong.

Reply:
Sure, just ignore anything I said to help you and spare yourself any kind of research. Hmpf.

```
Two expression evaluations conflict if one of them modifies a memory location
([intro.memory]) and the other one reads or modifies the same memory location.
[..] Any such data race results in undefined behavior
```

http://eel.is/c++draft/intro.multithread#intro.races-2

Reply:
OK, after some web searching (https://stackoverflow.com/questions/2171681/do-i-need-to-lock-object-when-reading-from-it), I found myself wrong. Reading technically also needs locking because the memory location might be in an intermediate state of writing. Although I doubt such an intermediate state could occur with `m_should_use_bootstrap_daemon` which is just a boolean, I'll add lock here as you suggest.

Reply:
Yeah, Booleans have no intermediate states, otherwise they would no longer be booleans. Data items of word-size or smaller are all atomic, you can never see only part of their bits being changed by an operation.

Reply:
@hyc that's not correct. It depends on the assembler whether the operation ends up being atomic or not. Secondly, without using a proper locking mechanism, there's still cache coherence where one thread sees the updated value but others don't. This is something we clearly need to avoid here, once set to `false` our value must not switch back to `true`.

See this discussion: https://stackoverflow.com/questions/16320838/when-do-i-really-need-to-use-atomicbool-instead-of-bool